WordDistanceInSpark.scala

```diff
@@ -1 +1,241 @@
+import org.apache.spark.SparkContext
+import org.apache.spark.SparkContext._
+import org.apache.spark.SparkConf
+import org.apache.spark.rdd.RDD
+
+
+object WordDistanceInSpark {
+  // I'm obviously doing something very wrong, because I get the following exception at the end on the run:
+/* ERROR ContextCleaner: Error in cleaning thread
+java.lang.InterruptedException
+  at java.lang.Object.wait(Native Method)
+  at java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:143)
+  at org.apache.spark.ContextCleaner$$anonfun$org$apache$spark$ContextCleaner$$keepCleaning$1.apply$mcV$sp(ContextCleaner.scala:146)
+  at org.apache.spark.ContextCleaner$$anonfun$org$apache$spark$ContextCleaner$$keepCleaning$1.apply(ContextCleaner.scala:144)
+  at org.apache.spark.ContextCleaner$$anonfun$org$apache$spark$ContextCleaner$$keepCleaning$1.apply(ContextCleaner.scala:144)
+  at org.apache.spark.util.Utils$.logUncaughtExceptions(Utils.scala:1618)
+  at org.apache.spark.ContextCleaner.org$apache$spark$ContextCleaner$$keepCleaning(ContextCleaner.scala:143)
+  at org.apache.spark.ContextCleaner$$anon$3.run(ContextCleaner.scala:65)
+ERROR Utils: Uncaught exception in thread SparkListenerBus
+java.lang.InterruptedException
+  at java.util.concurrent.locks.AbstractQueuedSynchronizer.doAcquireSharedInterruptibly(AbstractQueuedSynchronizer.java:998)
+  at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquireSharedInterruptibly(AbstractQueuedSynchronizer.java:1304)
+  at java.util.concurrent.Semaphore.acquire(Semaphore.java:312)
+  at org.apache.spark.util.AsynchronousListenerBus$$anon$1$$anonfun$run$1.apply$mcV$sp(AsynchronousListenerBus.scala:62)
+  at org.apache.spark.util.AsynchronousListenerBus$$anon$1$$anonfun$run$1.apply(AsynchronousListenerBus.scala:61)
+  at org.apache.spark.util.AsynchronousListenerBus$$anon$1$$anonfun$run$1.apply(AsynchronousListenerBus.scala:61)
+  at org.apache.spark.util.Utils$.logUncaughtExceptions(Utils.scala:1618)
+  at org.apache.spark.util.AsynchronousListenerBus$$anon$1.run(AsynchronousListenerBus.scala:60)*/
+
+
+  
+  
+  
+  type Hit = (Long, String) // (index, the significant word found at index)
+  type SameWordInterval = ((Long,Long),String) // ( (startIdx, endIdx), significant word)
```

Comment:
A Hit is a word at a certain index in the text. Specifically, one of the two words we care about.
I had noticed that the significant words occur in an order that is far from strictly interleaved.
E.g. instead of ... w1 ... w2 ... w1 ... w1 ... w2 ... w1 ... w2 ... w1
There are lactually ong sequences of the same significant word occurring before the other one occurs.
Since we only care about transitions between w1 and w2, only the first and last hit in a series of hits with the same word actually matter.
So that is why I condense the data down to knowing only that e.g. only significant word w1 is seen in the closed interval [firstIndexInSequence, lastIndexInSequence] .
The numbers show that for the first example of finding distance between Spark and version in the README.md that comes with spark-notebook, there are 36 hits of one of those two words, but they are collapsable into only 12 SameWordIntervals . For the second example of romeo and julia, there are 158 hits for either word, but the order in which they appear means that there are only 50 SameWordIntervals.
Admittedly, this whole focus for making computations on large data performant by shrinking the data as much as possible as soon as possible, comes straight from advice when I was still at university. It might be less relevant when a certain number of spark partitions are allocated anyway. I may also be overfitting to the specific cases I randomly choose for testing, but a factor of 3 for both tests looked like nothing to sniff at. And I must admit that the elegance  - of throwing away the datapoints on the insides of an extended interval of same words - appealed to me intuitively as well. Whether the extra computation to do this collapsing is worth it, ... well I did not test on a large dataset of texts on a cluster, so I just don't know.